[Docs][KubeRay] add example for distributed checkpointing with kuberay and gcsfuse (#45043)
ray-project/ray: doc/source/cluster/kubernetes/examples/distributed-checkpointing-with-gcsfuse.md

While it's possible to directly reference remote storage paths (e.g., `gs://my-checkpoint-bucket`), using Google Cloud Storage FUSE (GCSFuse) offers distinct advantages for distributed applications. GCSFuse allows you to mount Cloud Storage buckets like local file systems, making checkpoint management more intuitive for distributed applications that rely on these semantics. Furthermore, GCSFuse is designed for high-performance workloads, delivering the performance and scalability needed for distributed checkpointing of large models.

[Distributed checkpointing](https://docs.ray.io/en/latest/train/user-guides/checkpoints.html), in combination with [GCSFuse](https://cloud.google.com/storage/docs/gcs-fuse), allows for larger-scale model training with increased availability and efficiency.

## Create a Kubernetes cluster on GKE

Create a GKE cluster with the [GCSFuse CSI driver](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver) and [Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) enabled, as well as a GPU node pool with 4 L4 GPUs:
```

> export PROJECT_ID=<your project id>

Author: I think most users will know their PROJECT_ID because you need to create the GKE cluster. But PROJECT_NUMBER is kind of annoying to get. Unfortunately you need to know both. I added a reference to https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects to make it easier to find it

> gcloud container clusters create kuberay-with-gcsfuse \

Author: Not sure why it works for me. I added `--release-channel rapid` for now to ensure the correct release channel is used that supports this version

> gcloud container clusters create kuberay-with-gcsfuse \

Reviewer: It still fails, but I can now see the release newer than 1.29.3.

```
Default change: During creation of nodepools or autoscaling configuration changes for cluster versions greater than 1.24.1-gke.800 a default location policy is applied. For Spot and PVM it defaults to ANY, and for all other VM kinds a BALANCED policy is used. To change the default values use the `--location-policy` flag.
Note: Your Pod address range (`--cluster-ipv4-cidr`) can accommodate at most 1008 node(s).
Note: Machines with GPUs have certain limitations which may affect your workflow. Learn more at https://cloud.google.com/kubernetes-engine/docs/how-to/gpus
ERROR: (gcloud.container.clusters.create) ResponseError: code=400, message=Master version must be one of "RAPID" channel supported versions [1.26.15-gke.1191000, 1.26.15-gke.1300000, 1.27.13-gke.1000000, 1.27.13-gke.1166000, 1.28.9-gke.1000000, 1.28.9-gke.1209000, 1.29.3-gke.1282001, 1.29.4-gke.1043000, 1.29.4-gke.1447000, 1.30.0-gke.1167000, 1.30.0-gke.1457000].
```

> gcloud container clusters create kuberay-with-gcsfuse \

Reviewer: The command that I use successfully creates a Kubernetes cluster.

```sh
gcloud container clusters create kuberay-with-gcsfuse \
  --addons GcsFuseCsiDriver \
  --cluster-version=1.29.4 \
  --location=us-east4-c \
  --machine-type=g2-standard-8 \
  --release-channel=rapid \
  --num-nodes=4 \
  --accelerator type=nvidia-l4,count=1,gpu-driver-version=late
```

> gcloud container clusters create kuberay-with-gcsfuse \

Author: Let's use 1.29.4

    --addons GcsFuseCsiDriver \
    --cluster-version=1.29.3-gke.1093000 \
    --location=us-east4-c \
    --machine-type=g2-standard-8 \
    --num-nodes=4 \
    --accelerator type=nvidia-l4,count=1,gpu-driver-version=latest \
    --workload-pool=${PROJECT_ID}.svc.id.goog
```

## Install the KubeRay operator

Follow [Deploy a KubeRay operator](kuberay-operator-deploy) to install the latest stable KubeRay operator from the Helm repository.

> The KubeRay operator Pod must be on the CPU node if you set up the taint for the GPU node pool correctly.

Author: GKE adds taints, but the example has a toleration already https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/pytorch-resnet-image-classifier/ray-job.pytorch-image-classifier.yaml#L82-L85

## Configuring the GCS Bucket

Create a GCS bucket that will be used as the remote filesystem.
```

> gcloud storage buckets create gs://<your bucket name>

Author: Good idea, I added it

```

Create a Kubernetes ServiceAccount that will be used to grant the RayCluster access to mount the GCS bucket:
```
kubectl create serviceaccount pytorch-distributed-training
```

> Bind the `roles/storage.objectUser` role to the Kubernetes service account and bucket IAM policy:

Author: The output of the `gcloud storage buckets add-iam-policy-binding` command should tell you if the binding was successful. Otherwise you won't know until you try to access the bucket with the service account.

```
PROJECT_ID=<your project ID>
PROJECT_NUMBER=<your project number>

> BUCKET=<your GCS BUCKET>

Author: yes

```

See [Access Cloud Storage buckets with the Cloud Storage FUSE CSI driver](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver) for more details.

## Deploy the RayJob

Download the RayJob that executes all the steps documented in [Finetuning a Pytorch Image Classifier with Ray Train](https://docs.ray.io/en/latest/train/examples/pytorch/pytorch_resnet_finetune.html). The [source code](https://github.com/ray-project/kuberay/tree/master/ray-operator/config/samples/pytorch-resnet-image-classifier) is also in the KubeRay repository.

```
curl -LO https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/pytorch-resnet-image-classifier/ray-job.pytorch-image-classifier.yaml
```

Modify the RayJob by replacing all instances of the `GCS_BUCKET` placeholder with the Google Cloud Storage bucket you created earlier. Alternatively you can use `sed`:
```
sed -i "s/GCS_BUCKET/<your-bucket>/g" ray-job.pytorch-image-classifier.yaml
```

Deploy the RayJob
```

> kubectl create -f ray-job.pytorch-image-classifier.yaml

Reviewer: The worker's init container can't complete.

```sh
> kubectl describe pod lassifier-h4h2b-raycluster-ldzg2-worker-gpu-group-q6zxf

Events:
  Type     Reason     Age               From               Message
  ----     ------     ----              ----               -------
  Normal   Scheduled  8m4s              default-scheduler  Successfully assigned default/lassifier-h4h2b-raycluster-ldzg2-worker-gpu-group-q6zxf to gke-kuberay-with-gcsfuse-default-pool-845e7d9f-mhhv
  Normal   Pulling    8m3s              kubelet            Pulling image "rayproject/ray-ml:2.9.0-gpu"
  Normal   Pulled     3m27s             kubelet            Successfully pulled image "rayproject/ray-ml:2.9.0-gpu" in 4m36.355s (4m36.355s including waiting)
  Warning  Failed     87s               kubelet            Error: context deadline exceeded
  Normal   Pulled     2s (x8 over 87s)  kubelet            Container image "rayproject/ray-ml:2.9.0-gpu" already present on machine
  Warning  Failed     2s (x8 over 87s)  kubelet            Error: failed to reserve container name "wait-gcs-ready_lassifier-h4h2b-raycluster-ldzg2-worker-gpu-group-q6zxf_default_fb1d16f1-8425-4a73-8b83-2bbca5eebc19_0": name "wait-gcs-ready_lassifier-h4h2b-raycluster-ldzg2-worker-gpu-group-q6zxf_default_fb1d16f1-8425-4a73-8b83-2bbca5eebc19_0" is reserved for "a586844444d34eedee90c6efa2cdac54411e0eb711a5c721993a252697a88b15"

> kubectl logs lassifier-h4h2b-raycluster-ldzg2-worker-gpu-group-q6zxf -c wait-gcs-ready
Error from server (BadRequest): container "wait-gcs-ready" in pod "lassifier-h4h2b-raycluster-ldzg2-worker-gpu-group-q6zxf" is waiting to start: CreateContainerError
```

Author: The init container bug is related to https://github.com/ray-project/kuberay/issues/1130, this is why the example references a newer v1.29.3 GKE version . If you created with 1.29.1, it won't have the fix mentioned in the issue

Author: @kevin85421 I updated the gcloud command above to use `--release-channel=rapid` so you should be able to specify GKE version 1.29.3-gke.1093000 which fixes this issue. Can you try again with the updated version?

```

> The deployed RayJob includes the following configuration to enable distributed checkpointing to a shared filesystem:

Reviewer: Would you mind adding more details about the following configurations in the YAML? Currently, users can't understand how the storage looks.

```yaml
annotations:
  gke-gcsfuse/volumes: "true"
  gke-gcsfuse/cpu-limit: "0"
  gke-gcsfuse/memory-limit: 5Gi
  gke-gcsfuse/ephemeral-storage-limit: 10Gi
...
csi:
  driver: gcsfuse.csi.storage.gke.io
  volumeAttributes:
    bucketName: GCS_BUCKET
    mountOptions: "implicit-dirs,uid=1000,gid=100"
```

Author: I don't want to go into too much detail about all the possible annotations and mount options. I added this with some links to GCP docs:
```
The GCSFuse volume is configured with annotations which allows for finer grain control of the GCSFuse sidecar container. See [Specify Pod annotations](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#pod-annotations) for more details.
```
annotations:
  gke-gcsfuse/volumes: "true"
  gke-gcsfuse/cpu-limit: "0"
  gke-gcsfuse/memory-limit: 5Gi
  gke-gcsfuse/ephemeral-storage-limit: 10Gi
```

Mount options are also specified when defining the GCSFuse container volume:
```
csi:
  driver: gcsfuse.csi.storage.gke.io
  volumeAttributes:
    bucketName: GCS_BUCKET
    mountOptions: "implicit-dirs,uid=1000,gid=100"
```

See [Mount options](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#mount-options) to learn more about mount options.
```

[Distributed checkpointing](https://docs.ray.io/en/latest/train/user-guides/checkpoints.html), in combination with [GCSFuse](https://cloud.google.com/storage/docs/gcs-fuse), allows for larger-scale model training with increased availability and efficiency.

## Create a Kubernetes cluster on GKE

Create a GKE cluster with the [GCSFuse CSI driver](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver) and [Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) enabled, as well as a GPU node pool with 4 L4 GPUs:
```
export PROJECT_ID=<your project id>
gcloud container clusters create kuberay-with-gcsfuse \
    --addons GcsFuseCsiDriver \
    --cluster-version=1.29.3-gke.1093000 \
    --location=us-east4-c \
    --machine-type=g2-standard-8 \
    --release-channel=rapid \
    --num-nodes=4 \
    --accelerator type=nvidia-l4,count=1,gpu-driver-version=latest \
    --workload-pool=${PROJECT_ID}.svc.id.goog

> ```

Reviewer: Could we add some instructions to help users verify whether the GKE cluster is created successfully? For example,

```
> kubectl get nodes "-o=custom-columns=NAME:.metadata.name,GPU:.status.allocatable.nvidia\.com/gpu"
NAME                                                  GPU
gke-kuberay-with-gcsfuse-default-pool-xxxx-0000   1
gke-kuberay-with-gcsfuse-default-pool-xxxx-1111   1
gke-kuberay-with-gcsfuse-default-pool-xxxx-2222   1
gke-kuberay-with-gcsfuse-default-pool-xxxx-3333   1
```

Author: Good idea, I added it

## Install the KubeRay operator

Follow [Deploy a KubeRay operator](kuberay-operator-deploy) to install the latest stable KubeRay operator from the Helm repository.
The KubeRay operator Pod must be on the CPU node if you set up the taint for the GPU node pool correctly.

## Configuring the GCS Bucket

Create a GCS bucket that will be used as the remote filesystem.
```

> BUCKET=<your GCS buckert>

Reviewer: Update: GCS_BUCKET is a invalid GCS bucket name. It is just a placeholder in the YAML.

> BUCKET=<your GCS buckert>

Author: yes, there's a step below to replace the placeholder

> BUCKET=<your GCS buckert>

Author: fixed

(kuberay-distributed-checkpointing-gcsefuse)=

> # Distributed Checkpointing with KubeRay and GCSFuse

Reviewer 2: ```suggestion
# Distributed checkpointing with KubeRay and GCSFuse
```

> This example demonstrates how to orchestrate distributed checkpointing with KubeRay, using the GCSFuse CSI driver and Google Cloud Storage as the remote storage system.

Reviewer 2: ```suggestion
This example orchestrates distributed checkpointing with KubeRay, using the GCSFuse CSI driver and Google Cloud Storage as the remote storage system.
```

To illustrate the concepts, this guide uses the [Finetuning a Pytorch Image Classifier with Ray Train](https://docs.ray.io/en/latest/train/examples/pytorch/pytorch_resnet_finetune.html) example.

> ## Why Distributed Checkpointing with GCSFuse?

Reviewer 2: ```suggestion
## Distributed checkpointing with GCSFuse?
```

> In large-scale, high-performance machine learning, distributed checkpointing is crucial for fault tolerance, ensuring that if a node fails during training, the process can be resumed from the latest saved checkpoint instead of starting from scratch.

Reviewer 2: ```suggestion
In large-scale, high-performance machine learning, distributed checkpointing is crucial for fault tolerance, ensuring that if a node fails during training, Ray can resume the process from the latest saved checkpoint instead of starting from scratch.
```

> While it's possible to directly reference remote storage paths (e.g., `gs://my-checkpoint-bucket`), using Google Cloud Storage FUSE (GCSFuse) offers distinct advantages for distributed applications. GCSFuse allows you to mount Cloud Storage buckets like local file systems, making checkpoint management more intuitive for distributed applications that rely on these semantics. Furthermore, GCSFuse is designed for high-performance workloads, delivering the performance and scalability needed for distributed checkpointing of large models.

Reviewer 2: ```suggestion
While it's possible to directly reference remote storage paths (e.g., `gs://my-checkpoint-bucket`), using Google Cloud Storage FUSE (GCSFuse) has distinct advantages for distributed applications. GCSFuse allows you to mount Cloud Storage buckets like local file systems, making checkpoint management more intuitive for distributed applications that rely on these semantics. Furthermore, GCSFuse is designed for high-performance workloads, delivering the performance and scalability you need for distributed checkpointing of large models.
```

```
export PROJECT_ID=<your project id>
gcloud container clusters create kuberay-with-gcsfuse \
    --addons GcsFuseCsiDriver \
    --cluster-version=1.29.4 \
    --location=us-east4-c \
    --machine-type=g2-standard-8 \
    --release-channel=rapid \
    --num-nodes=4 \
    --accelerator type=nvidia-l4,count=1,gpu-driver-version=latest \
    --workload-pool=${PROJECT_ID}.svc.id.goog
```

Verify the successful creation of your cluster with 4 GPUs:
```
$ kubectl get nodes "-o=custom-columns=NAME:.metadata.name,GPU:.status.allocatable.nvidia\.com/gpu"
NAME                                                  GPU
gke-kuberay-with-gcsfuse-default-pool-xxxx-0000       1
gke-kuberay-with-gcsfuse-default-pool-xxxx-1111       1
gke-kuberay-with-gcsfuse-default-pool-xxxx-2222       1
gke-kuberay-with-gcsfuse-default-pool-xxxx-3333       1
```

## Install the KubeRay operator

Follow [Deploy a KubeRay operator](kuberay-operator-deploy) to install the latest stable KubeRay operator from the Helm repository.
The KubeRay operator Pod must be on the CPU node if you set up the taint for the GPU node pool correctly.

## Configuring the GCS Bucket

> Create a GCS bucket that will be used as the remote filesystem.

Reviewer 2: ```suggestion
Create a GCS bucket that KubeRay uses as the remote filesystem.
```

Reviewer 2: Not sure if it's KubeRay that uses the remote filesystem, but trying to be more precise on who is using it. It's ambiguous with the passive voice. Please correct the subject as needed.

```
BUCKET=<your GCS bucket>
gcloud storage buckets create gs://$BUCKET --uniform-bucket-level-access
```

> Create a Kubernetes ServiceAccount that will be used to grant the RayCluster access to mount the GCS bucket:

Reviewer 2: ```suggestion
Create a Kubernetes ServiceAccount that KubeRay uses to grant the RayCluster access to mount the GCS bucket:
```

Reviewer 2: We follow the Google Developer Documentation Style Guide (https://developers.google.com/style), which discourages the use of future tense.

```
kubectl create serviceaccount pytorch-distributed-training
```

Bind the `roles/storage.objectUser` role to the Kubernetes service account and bucket IAM policy.

> See [identifying projects](https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects) to figure out your project ID and project number:

Reviewer 2: ```suggestion
See [Identifying projects](https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects) to find your project ID and project number:
```

```

See [Access Cloud Storage buckets with the Cloud Storage FUSE CSI driver](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver) for more details.

## Deploy the RayJob

Download the RayJob that executes all the steps documented in [Finetuning a Pytorch Image Classifier with Ray Train](https://docs.ray.io/en/latest/train/examples/pytorch/pytorch_resnet_finetune.html). The [source code](https://github.com/ray-project/kuberay/tree/master/ray-operator/config/samples/pytorch-resnet-image-classifier) is also in the KubeRay repository.

```
curl -LO https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/pytorch-resnet-image-classifier/ray-job.pytorch-image-classifier.yaml
```

Modify the RayJob by replacing all instances of the `GCS_BUCKET` placeholder with the Google Cloud Storage bucket you created earlier. Alternatively you can use `sed`:
```
sed -i "s/GCS_BUCKET/$BUCKET/g" ray-job.pytorch-image-classifier.yaml
```

> Deploy the RayJob

Reviewer 2: ```suggestion
Deploy the RayJob:
```

```
kubectl create -f ray-job.pytorch-image-classifier.yaml
```

The deployed RayJob includes the following configuration to enable distributed checkpointing to a shared filesystem:
* 4 Ray workers, each with a single GPU.

> * All Ray nodes use the `pytorch-distributed-training` ServiceAccount which we created earlier.

Reviewer 2: ```suggestion
* All Ray nodes use the `pytorch-distributed-training` ServiceAccount, which we created earlier.
```

* Includes volumes that are managed by the `gcsfuse.csi.storage.gke.io` CSI driver.
* Mounts a shared storage path `/mnt/cluster_storage`, backed by the GCS bucket you created earlier.

> The Pod can be configured with annotations which allows for finer grain control of the GCSFuse sidecar container. See [Specify Pod annotations](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#pod-annotations) for more details.

Reviewer 2: ```suggestion
You can configure the Pod with annotations, which allows for finer grain control of the GCSFuse sidecar container. See [Specify Pod annotations](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#pod-annotations) for more details.
```

```
annotations:
  gke-gcsfuse/volumes: "true"
  gke-gcsfuse/cpu-limit: "0"
  gke-gcsfuse/memory-limit: 5Gi
  gke-gcsfuse/ephemeral-storage-limit: 10Gi
```

> Mount options are also specified when defining the GCSFuse container volume:

Reviewer 2: ```suggestion
You also specify mount options when defining the GCSFuse container volume:
```

```
csi:
  driver: gcsfuse.csi.storage.gke.io
  volumeAttributes:
    bucketName: GCS_BUCKET
    mountOptions: "implicit-dirs,uid=1000,gid=100"
```

See [Mount options](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#mount-options) to learn more about mount options.

> Logs from the Ray job should indicate use of the shared remote filesystem in `/mnt/cluster_storage` and the checkpointing directory. For example:

Reviewer 2: ```suggestion
Logs from the Ray job should indicate the use of the shared remote filesystem in `/mnt/cluster_storage` and the checkpointing directory. For example:
```

gs://my-ray-bucket/finetune-resnet/TorchTrainer_96923_00000_0_2024-04-29_17-21-29/events.out.tfevents.1714436502.orch-image-classifier-nc2sq-raycluster-tdrfx-head-xzcl8
gs://my-ray-bucket/finetune-resnet/TorchTrainer_96923_00000_0_2024-04-29_17-21-29/events.out.tfevents.1714436809.orch-image-classifier-zz4sj-raycluster-vn7kz-head-lwx8k
gs://my-ray-bucket/finetune-resnet/TorchTrainer_96923_00000_0_2024-04-29_17-21-29/params.json
gs://my-ray-bucket/finetune-resnet/TorchTrainer_96923_00000_0_2024-04-29_17-21-29/params.pkl
gs://my-ray-bucket/finetune-resnet/TorchTrainer_96923_00000_0_2024-04-29_17-21-29/progress.csv
gs://my-ray-bucket/finetune-resnet/TorchTrainer_96923_00000_0_2024-04-29_17-21-29/result.json
gs://my-ray-bucket/finetune-resnet/basic-variant-state-2024-04-29_17-21-29.json
gs://my-ray-bucket/finetune-resnet/basic-variant-state-2024-04-29_17-26-35.json
gs://my-ray-bucket/finetune-resnet/experiment_state-2024-04-29_17-21-29.json
gs://my-ray-bucket/finetune-resnet/experiment_state-2024-04-29_17-26-35.json
gs://my-ray-bucket/finetune-resnet/trainer.pkl
gs://my-ray-bucket/finetune-resnet/tuner.pkl

```

## Resuming from checkpoint

> In the event of a failed job, the latest checkpoint can be used to resume training our model. The `TorchTrainer` in our example is configured to automatically resume

Reviewer 2: ```suggestion
In the event of a failed job, you can use the latest checkpoint to resume training of the model. This example configures `TorchTrainer` to automatically resume
```

```
2024-04-29 15:51:32,528 INFO experiment_state.py:366 -- Trying to find and download experiment checkpoint at /mnt/cluster_storage/finetune-resnet
2024-04-29 15:51:32,651 INFO experiment_state.py:396 -- A remote experiment checkpoint was found and will be used to restore the previous experiment state.
2024-04-29 15:51:32,652 INFO tune_controller.py:404 -- Using the newest experiment state file found within the experiment directory: experiment_state-2024-04-29_15-43-40.json

View detailed results here: /mnt/cluster_storage/finetune-resnet
To visualize your results with TensorBoard, run: `tensorboard --logdir /home/ray/ray_results/finetune-resnet`

Result(
  metrics={'loss': 0.070047477101968, 'acc': 0.23529411764705882},
  path='/mnt/cluster_storage/finetune-resnet/TorchTrainer_ecc04_00000_0_2024-04-29_15-43-40',
  filesystem='local',
  checkpoint=Checkpoint(filesystem=local, path=/mnt/cluster_storage/finetune-resnet/TorchTrainer_ecc04_00000_0_2024-04-29_15-43-40/checkpoint_000009)
)
```

If the previous job failed at an earlier checkpoint, the job should resume from the last saved checkpoint and run until `max_epochs=10`. For example, if the last run

> failed at epoch 7, the training will automatically resume using `checkpoint_000006` and run 3 more iterations until epoch 10:

Reviewer 2: ```suggestion
failed at epoch 7, the training automatically resumes using `checkpoint_000006` and run 3 more iterations until epoch 10:
```

gke-kuberay-with-gcsfuse-default-pool-xxxx-0000       1
gke-kuberay-with-gcsfuse-default-pool-xxxx-1111       1
gke-kuberay-with-gcsfuse-default-pool-xxxx-2222       1
gke-kuberay-with-gcsfuse-default-pool-xxxx-3333       1
```

## Install the KubeRay operator

Follow [Deploy a KubeRay operator](kuberay-operator-deploy) to install the latest stable KubeRay operator from the Helm repository.
The KubeRay operator Pod must be on the CPU node if you set up the taint for the GPU node pool correctly.

## Configuring the GCS Bucket

Create a GCS bucket that will be used as the remote filesystem.
```
BUCKET=<your GCS bucket>
gcloud storage buckets create gs://$BUCKET
```

Create a Kubernetes ServiceAccount that will be used to grant the RayCluster access to mount the GCS bucket:
```
kubectl create serviceaccount pytorch-distributed-training
```

Bind the `roles/storage.objectUser` role to the Kubernetes service account and bucket IAM policy.
See [identifying projects](https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects) to figure out your project ID and project number:
```
PROJECT_ID=<your project ID>
PROJECT_NUMBER=<your project number>

> BUCKET=<your GCS BUCKET>

Reviewer: Maybe we can remove this one since we have already set the env above.

Author: done

```

See [Access Cloud Storage buckets with the Cloud Storage FUSE CSI driver](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver) for more details.

## Deploy the RayJob

Download the RayJob that executes all the steps documented in [Finetuning a Pytorch Image Classifier with Ray Train](https://docs.ray.io/en/latest/train/examples/pytorch/pytorch_resnet_finetune.html). The [source code](https://github.com/ray-project/kuberay/tree/master/ray-operator/config/samples/pytorch-resnet-image-classifier) is also in the KubeRay repository.

```
curl -LO https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/pytorch-resnet-image-classifier/ray-job.pytorch-image-classifier.yaml
```

Modify the RayJob by replacing all instances of the `GCS_BUCKET` placeholder with the Google Cloud Storage bucket you created earlier. Alternatively you can use `sed`:
```

> sed -i "s/GCS_BUCKET/<your GCS bucket>/g" ray-job.pytorch-image-classifier.yaml

Reviewer: ```suggestion
sed -i "s/GCS_BUCKET/$BUCKET/g" ray-job.pytorch-image-classifier.yaml
```

Author: dome

* Includes volumes that are managed by the `gcsfuse.csi.storage.gke.io` CSI driver.
* Mounts a shared storage path `/mnt/cluster_storage`, backed by the GCS bucket you created earlier.

The Pod can be configured with annotations which allows for finer grain control of the GCSFuse sidecar container. See [Specify Pod annotations](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#pod-annotations) for more details.
```
annotations:
  gke-gcsfuse/volumes: "true"
  gke-gcsfuse/cpu-limit: "0"
  gke-gcsfuse/memory-limit: 5Gi
  gke-gcsfuse/ephemeral-storage-limit: 10Gi
```

Mount options are also specified when defining the GCSFuse container volume:
```
csi:
  driver: gcsfuse.csi.storage.gke.io
  volumeAttributes:
    bucketName: GCS_BUCKET
    mountOptions: "implicit-dirs,uid=1000,gid=100"
```

See [Mount options](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#mount-options) to learn more about mount options.

Logs from the Ray job should indicate use of the shared remote filesystem in `/mnt/cluster_storage` and the checkpointing directory. For example:
```

> Training finished iteration 10 at 2024-04-29 10:22:08. Total running time: 1min 30s

Reviewer: The job fails; I will troubleshoot the issue.

<img width="1440" alt="Screenshot 2024-05-17 at 1 18 38 PM" src="https://github.com/ray-project/ray/assets/20109646/749a0397-28ef-45ce-99c6-783cc3babf15">

Reviewer: I didn’t have a chance to dive deep into troubleshooting this issue, but it seems to be a GCSFuse bug (or a lack of observability). I tried to create a file in `/mnt/cluster_storage`, and although the command didn't report an error, I couldn’t see the file.

![Screenshot 2024-05-17 at 2 00 29 PM](https://github.com/ray-project/ray/assets/20109646/a071cef8-b07b-49c5-bf19-1aa9ef374bae)

Reviewer: Update: Synced with @andrewsykim offline. We need to set the GCS bucket's permissions to 'Uniform' instead of fine-grained.
